### PR TITLE
Add NetFlow probe

### PR DESCRIPTION
**Pull request by safchain**

### CODE REVIEW

File: api/types/types.go [hunk 1]
```diff
@@ -145,10 +146,10 @@ type PacketInjection struct {
 	BasicResource    `yaml:",inline"`
 	Src              string `yaml:"Src"`
 	Dst              string `yaml:"Dst"`
-	SrcIP            string `yaml:"SrcIP"`
-	DstIP            string `yaml:"DstIP"`
-	SrcMAC           string `yaml:"SrcMAC"`
-	DstMAC           string `yaml:"DstMAC"`
+	SrcIP            string `valid:"isCIDR" yaml:"SrcIP"`
```
Comment (hunchback) [inline api/types/types.go]: `valid:"isIP"` and then use `new.ParseIP()` in implementation

File: api/types/types.go [hunk 2]
```diff
@@ -145,10 +146,10 @@ type PacketInjection struct {
...
+	DstIP            string `valid:"isCIDR" yaml:"DstIP"`
+	SrcMAC           string `valid:"isMAC" yaml:"SrcMAC"`
+	DstMAC           string `valid:"isMAC" yaml:"DstMAC"`
 	SrcPort          int64  `yaml:"SrcPort"`
```
Comment (hunchback) [inline api/types/types.go]: You may also consider validating port range - you can take some ideas from: https://golang.org/src/net/port.go

File: flow/probes/ovssflow.go [hunk 3]
```diff
@@ -173,50 +176,41 @@ func (o *OvsSFlowProbesHandler) UnregisterSFlowProbeFromBridge(bridgeUUID string
...
-	if err != nil && err != sflow.ErrAgentAlreadyAllocated {
-		return err
-	}
+	if capture.Target == "" {
+		opts := tableOptsFromCapture(capture)
+		probe.flowTable = o.fpta.Alloc(tid, opts)
 
-	probe.Target = agent.GetTarget()
+		address := config.GetString("sflow.bind_address")
+		if address == "" {
```
Comment (hunchback) [inline flow/probes/ovssflow.go]: should default not be handled in `config/config.go` like most other parameters?

File: netflow/agent.go [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+	a.agents = append(a.agents, s)
+
+	s.Start()
+	return s, nil
+}
+
+// NewAgentAllocator creates a new sFlow agent allocator
+func NewAgentAllocator() (*AgentAllocator, error) {
+	min := config.GetInt("netflow.port_min")
+	max := config.GetInt("netflow.port_max")
```
Comment (hunchback) [inline netflow/agent.go]: verify that max >= min

File: flow/probes/ovssflow.go (hunk 3 above)
Comment (safchain) [reply flow/probes/ovssflow.go]: not really, if the user provide an empty value in the config file default won't be applied

File: netflow/agent.go (hunk 4 above)
Comment (safchain) [reply netflow/agent.go]: checked by the next line

**Comment by safchain** [comment]
run skydive-functional-tests-backend-orientdb

**Comment by safchain** [comment]
run skydive-functional-tests-backend-orientdb

**Comment by safchain** [comment]
run skydive-functional-tests-backend-orientdb

**Comment by safchain** [comment]
run skydive-functional-tests-backend-orientdb

**Comment by safchain** [comment]
run skydive-functional-tests-backend-orientdb

**Comment by safchain** [comment]
run skydive-cdd-overview-tests

**Comment by safchain** [comment]
run skydive-k8s-tests

File: flow/probes/ovsnetflow.go [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+	condition := libovsdb.NewCondition("_uuid", "==", libovsdb.UUID{GoUUID: bridgeUUID})
+	updateOp := libovsdb.Operation{
+		Op:    "update",
+		Table: "Bridge",
+		Row:   bridgeRow,
+		Where: []interface{}{condition},
+	}
+
+	operations = append(operations, updateOp)
+	_, err = o.ovsClient.Exec(operations...)
```
Comment (lebauce) [inline flow/probes/ovsnetflow.go]: return err

File: flow/probes/ovsnetflow.go [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+
+	if probe.flowTable != nil {
+		o.fpta.Release(probe.flowTable)
+	}
+	o.probesLock.RUnlock()
+
+	probeUUID, err := ovsRetrieveSkydiveProbeRowUUID(o.ovsClient, "NetFlow", ovsProbeID(bridgeUUID))
+	if err != nil {
+		return err
+	}
```
Comment (lebauce) [inline flow/probes/ovsnetflow.go]: else if probeUUID == "" {

File: flow/probes/ovsnetflow.go [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+
+	extIds := make(map[string]string)
+	extIds["skydive-probe-id"] = ovsProbeID(probe.ID)
+	ovsMap, err := libovsdb.NewOvsMap(extIds)
+	if err != nil {
+		return nil, err
+	}
+	netFlowRow["external_ids"] = ovsMap
+
+	insertOp := libovsdb.Operation{
```
Comment (lebauce) [inline flow/probes/ovsnetflow.go]: return libovsdb.Operation{...}, nil

File: flow/probes/pcapsocket.go [hunk 8]
```diff
@@ -89,18 +89,22 @@ func (p *PcapSocketProbeHandler) registerProbe(n *graph.Node, capture *types.Cap
...
-	}
-
+	var listener *net.TCPListener
+	var err error
 	var tcpAddr = *p.addr
-	tcpAddr.Port = port
+	fnc := func(p int) error {
+		listener, err = net.ListenTCP("tcp", &tcpAddr)
+		if err != nil {
+
```
Comment (lebauce) [inline flow/probes/pcapsocket.go]: missing `return err`

File: netflow/agent.go [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+	common.RWMutex
+	portAllocator *common.PortAllocator
+	agents        []*Agent
+}
+
+// GetTarget returns the current used connection
+func (nfa *Agent) GetTarget() string {
+	return fmt.Sprintf("%s:%d", nfa.Addr, nfa.Port)
+}
+func intToIP(nn uint32) net.IP {
```
Comment (lebauce) [inline netflow/agent.go]: nit: missing empty line

File: netflow/agent.go [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+
+// Start the NetFlow probe agent
+func (nfa *Agent) Start() {
+	go nfa.start()
+}
+
+// Stop the NetFlow probe agent
+func (nfa *Agent) Stop() {
+	nfa.Lock()
+	defer nfa.Unlock()
```
Comment (lebauce) [inline netflow/agent.go]: Seems to do nothing

File: validator/validator.go [hunk 11]
```diff
@@ -42,7 +42,7 @@ var (
 
 	//IPNotValid validator
 	IPNotValid = func() error {
-		return valid.TextErr{Err: errors.New("Not a IP addr")}
+		return valid.TextErr{Err: errors.New("Not a IP address")}
```
Comment (lebauce) [inline validator/validator.go]: Not an IP address

File: flow/probes/ovssflow.go [hunk 12]
```diff
@@ -194,34 +193,29 @@ func (o *OvsSFlowProbesHandler) RegisterProbeOnBridge(bridgeUUID string, tid str
...
+			address = "127.0.0.1"
+		}
 
-	return o.registerSFlowProbeOnBridge(probe, bridgeUUID)
-}
+		addr := common.ServiceAddress{Addr: address, Port: 0}
```
Comment (lebauce) [inline flow/probes/ovssflow.go]: nit: remove Port: 0

File: cmd/client/capture.go [hunk 13]
```diff
@@ -194,6 +196,7 @@ func addCaptureFlags(cmd *cobra.Command) {
 	cmd.Flags().BoolVarP(&reassembleTCP, "reassamble-tcp", "", false, "Reassemble TCP packets, default: false")
 	cmd.Flags().StringVarP(&layerKeyMode, "layer-key-mode", "", "L2", "Defines the first layer used by flow key calculation, L2 or L3")
 	cmd.Flags().StringArrayVarP(&extraLayers, "extra-layer", "", []string{}, fmt.Sprintf("List of extra layers to be added to the flow, available: %s", flow.ExtraLayers(flow.ALLLayer)))
+	cmd.Flags().StringVarP(&target, "target", "", "", "Specify sFlow target, if empty the agent will be used")
```
Comment (lebauce) [inline cmd/client/capture.go]: sFlow/NetFlow

**Comment by safchain** [comment]
run skydive-functional-hw-tests
run skydive-functional-tests-backend-elasticsearch

**Comment by safchain** [comment]
run skydive-functional-hw-tests
run skydive-functional-tests-backend-elasticsearch

File: config/config.go [hunk 14]
```diff
@@ -56,6 +56,9 @@ var (
 		"analyzer.topology.backend":            {"graph.backend"},
 		"analyzer.topology.k8s.config_file":    {"k8s.config_file"},
 		"analyzer.topology.k8s.probes":         {"k8s.probes"},
+		"agent.topology.sflow.bind_address":    {"sflow.bind_address"},
```
Comment (lebauce) [inline config/config.go]: topology ?

File: flow/probes/ovsnetflow.go (hunk 5 above)
Comment (lebauce) [reply flow/probes/ovsnetflow.go]: missed this comment ?

File: config/config.go (hunk 14 above)
Comment (safchain) [reply config/config.go]: This one is doing both but better to move it in the flow section as rely on capture

File: flow/probes/ovsnetflow.go (hunk 5 above)
Comment (safchain) [reply flow/probes/ovsnetflow.go]: yep missed

**Comment by lebauce** [comment]
run skydive-k8s-tests

**Comment by safchain** [comment]
run skydive-functional-tests-backend-elasticsearch

File: etc/skydive.yml.default [hunk 15]
```diff
@@ -253,6 +253,51 @@ agent:
...
+    netflow:
+      # Default listening address is 127.0.0.1
+      # bind_address: 127.0.0.1
+
+      # Port min/max used when starting a netflow probe, an agent will be started
+      # with a port from this range
+      # port_min: 6365
+      # port_max: 6375
+
+  flow:
```
Comment (lebauce) [inline etc/skydive.yml.default]: This section is duplicated

File: flow/probes/ovsnetflow.go [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+	probe := OvsNetFlowProbe{
+		ID:       bridgeUUID,
+		EngineID: 1,
+	}
+
+	if capture.Target == "" {
+		opts := tableOptsFromCapture(capture)
+		probe.flowTable = o.fpta.Alloc(tid, opts)
+
+		address := config.GetString("agent.topology.netflow.bind_address")
```
Comment (lebauce) [inline flow/probes/ovsnetflow.go]: Should now be agent.flow.netflow.bind_address

File: flow/probes/ovssflow.go [hunk 17]
```diff
@@ -194,34 +193,29 @@ func (o *OvsSFlowProbesHandler) RegisterProbeOnBridge(bridgeUUID string, tid str
...
-	agent, err := o.allocator.Alloc(bridgeUUID, probe.flowTable, capture.BPFFilter, headerSize, &addr, n, o.Graph)
-	if err != nil && err != sflow.ErrAgentAlreadyAllocated {
-		return err
-	}
+	if capture.Target == "" {
+		opts := tableOptsFromCapture(capture)
+		probe.flowTable = o.fpta.Alloc(tid, opts)
 
-	probe.Target = agent.GetTarget()
+		address := config.GetString("sflow.bind_address")
```
Comment (lebauce) [inline flow/probes/ovssflow.go]: We should probably use agent.flow.sflow.bind_address now

**Pull request merged**